File: dsa/python_dsa/queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    def __init__(self, queue=None):
        if queue == None:
            self.items = []
        else:
            self.items = queue

    def enqueue(self, i):
        self.items.insert(0, i)

    def dequeue(self):
        return self.items.pop()

    def size(self):
        return len(self.items)

    def isEmpty(self):
        return self.size() == 0

    def __str__(self):
        return str(self.items)


class Linked_Queue:
    def __init__(self, queue=None):
        if queue == None:
            self.front = None
            self.rear = None
        else:
            self.front = queue.front
            self.rear = queue.rear

    def enqueue(self, i):
        temp = Node(i)
        if self.front == None:
            self.front = temp
            self.rear = temp
        else:
            self.rear.next = temp
            self.rear = temp

    def dequeue(self):
        if self.front == None:
            return None
        else:
            temp = self.front
            self.front = self.front.next
            return temp.data

    def size(self):
        count = 0
        p = self.front
        while p != None:
            count += 1
            p = p.next
        return count

    def isEmpty(self):
        return self.size() == 0

    def __str__(self):
        s = ""
        p = self.front
        while p != None:
            s += str(p.data) + " "
            p = p.next
        return s


class Arr_Queue:
    def __init__(self, n):
        self.items = [0]*n
        self.front = 0
        self.rear = n-1

    def enqueue(self, i):
        self.items.append(i)
        self.rear += 1

    def dequeue(self):
        self.front += 1

    def size(self):
        return self.rear - self.front + 1

    def isEmpty(self):
        return self.size() == 0

    def __str__(self):
        return str(self.items[self.front:self.rear+1])

File: dsa/Queue.cpp
#include <iostream>

using namespace std;

class Queue
{
protected:
    int *queue;
    int length;
    int size;
    short int front;
    short int rear;

public:
    Queue(int size)
    {
        this->length = 0;
        this->size = size;
        queue = new int[size];
        front = 0;
        rear = size - 1;
    }
    ~Queue();
    void enqueue(int data)
    {
        if (this->length == 0)
        {
            queue[front] = data;
            this->length++;
        }
        else if (this->length == this->size)
        {
            cout << "Queue is full" << endl;
        }
        else
        {
            rear = (rear + 1) % size;
            queue[rear] = data;
            this->length++;
        }
    }
    int dequeue()
    {
        if (this->length == 0)
        {
            cout << "Queue is empty" << endl;
            return -1;
        }
        else
        {
            int data = queue[front];
            front = (front + 1) % size;
            this->length--;
            return data;
        }
    }
    void print()
    {
        if (this->length == 0)
        {
            cout << "Queue is empty";
        }
        else
        {
            for (int i = 0; i < this->length; i++)
            {
                cout << queue[(front + i) % size] << " ";
            }
        }
    }
    bool isempty()
    {
        return this->length == 0;
    }
};

class dequeue : public Queue
{
public:
    dequeue(int size) : Queue(size) {}
    void enqueue_front(int data)
    {
        if (this->length == 0)
        {
            queue[front] = data;
            this->length++;
        }
        else if (this->length == this->size)
        {
            cout << "Queue is full" << endl;
        }
        else
        {
            front = (front - 1 + size) % size;
            queue[front] = data;
            this->length++;
        }
    }
    void enqueue_rear(int data)
    {
        if (this->length == 0)
        {
            queue[front] = data;
            this->length++;
        }
        else if (this->length == this->size)
        {
            cout << "Queue is full" << endl;
        }
        else
        {
            rear = (rear + 1) % size;
            queue[rear] = data;
            this->length++;
        }
    }
    int dequeue_front()
    {
        if (this->length == 0)
        {
            cout << "Queue is empty" << endl;
            return -1;
        }
        else
        {
            int data = queue[front];
            front = (front + 1) % size;
            this->length--;
            return data;
        }
    }
    int dequeue_rear()
    {
        if (this->length == 0)
        {
            cout << "Queue is empty" << endl;
            return -1;
        }
        else
        {
            int data = queue[rear];
            rear = (rear - 1 + size) % size;
            this->length--;
            return data;
        }
    }
};

class PriorityQueue : public Queue
{
public:
    PriorityQueue(int size) : Queue(size) {}
    void enqueue(int data)
    {
        if (this->length == 0)
        {
            queue[front] = data;
            this->length++;
        }
        else if (this->length == this->size)
        {
            cout << "Queue is full" << endl;
        }
        else
        {
            int i = this->length - 1;
            while (i >= 0 && queue[i] > data)
            {
                queue[i + 1] = queue[i];
                i--;
            }
            queue[i + 1] = data;
            this->length++;
        }
    }
};
